libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+                updateEventsQueue(telemetryEvent)
+            }
+        }
+    }
+
+    private fun updateEventsQueue(telemetryEvent: TelemetryEventInterface) {
+        telemetryEventsQueue.add(telemetryEvent)
+        if (telemetryEventsQueue.size >= MAX_TELEMTRY_EVENTS) {
+            telemetryEventsQueue.forEach { event ->
+                mapboxTelemetry.push(event as Event)
+            }
+            telemetryEventsQueue.clear()
+        }
+        telemetryEventsQueue.add(telemetryEvent)
+    }
+
+    private fun registerForNotification(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.registerRouteProgressObserver(routeProgressListener)
+    }
+
+    /**
+     * Monitors all channels used in this class. This is done so that access to the containers these channels act on is serialized.
+     * By serializing all calls through this choke point, we can guaranty that this class is thread safe.
+     */
+    private fun monitorChannels() {
+        cleanupJob = jobControl.scope.monitorChannelWithException(channelLocationBuffer) { locationAction ->
+            when (locationAction.command) {
+                LocationBufferCommands.BUFFER_ADD -> {
+                    ifNonNull(locationAction.location) { location ->
+                        locationBuffer.add(location)
+                        if (locationBuffer.size > MAX_LOCATION_VALUES) {
+                            locationBuffer.removeAt(0)
+                        }
+                    }
+                }
+                LocationBufferCommands.BUFFER_REMOVE -> {
+                    if (locationBuffer.size > 0) {
+                        locationBuffer.removeAt(0)
+                    }
+                }
+                LocationBufferCommands.BUFFER_CLEAR -> {
+                    locationBuffer.clear()
+                }
+                LocationBufferCommands.BUFFER_READ -> {
+                    locationBuffer.forEach { location ->
+                        locationAction.predicate(location)
+                    }
+                }
+            }
+        }
+
+        /**
+         * This monitors the life time of the job that handles the channels. If that job is terminated, we cleanup.
+         */
+        jobControl.scope.launch {
+            select<Unit> {
+                cleanupJob.onJoin {
+                    locationEngine.removeLocationUpdates(locationCallback)
+                }
+            }
+        }
+    }
+
+    private suspend fun getDatePartitionedLocations(predicate: (Location) -> Boolean): Array<Location> {
+        val retVal = mutableListOf<Location>()
+        channelLocationBuffer.send(LocationBufferControl(LocationBufferCommands.BUFFER_READ, null) { location ->
+            if (predicate(location)) {
+                retVal.add(location)
+            }
+        })
+        return retVal.toTypedArray()
+    }
+
+    private fun parseRouteProgressAsync(): Deferred<TelemetryStep> {
+        return jobControl.scope.async {
```

Comment:
```suggestion
        return withContext(IO) {
```

Reply:
See previous comment

Reply:
`withContext` also returns a value
just use 
```
private suspend fun parseRouteProgressAsync(): TelemetryStep {
        return withContext {...}
``` 
In your case you don't have multiple coroutines running in parallel, so no need in `async`.
Also with `async` you create additional coroutine object, and `withContext` doesn't do it, it just switches dispatchers and wait (suspend) for result

Reply:
👍